x/evm/types/ethtx/txdata.go

```diff
@@ -0,0 +1,92 @@
+package ethtx
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+var (
+	_ TxData = &LegacyTx{}
+	_ TxData = &AccessListTx{}
+	_ TxData = &DynamicFeeTx{}
+	_ TxData = &BlobTx{}
+)
+
+// Unfortunately `TxData` interface in go-ethereum/core/types defines its functions
+// as private, so we have to define our own here.
+type TxData interface {
+	TxType() byte
+	Copy() TxData
+	GetChainID() *big.Int
+	GetAccessList() ethtypes.AccessList
+	GetData() []byte
+	GetNonce() uint64
+	GetGas() uint64
+	GetGasPrice() *big.Int
+	GetGasTipCap() *big.Int
+	GetGasFeeCap() *big.Int
+	GetValue() *big.Int
+	GetTo() *common.Address
+
+	GetRawSignatureValues() (v, r, s *big.Int)
+	SetSignatureValues(chainID, v, r, s *big.Int)
+
+	AsEthereumData() ethtypes.TxData
+	Validate() error
+
+	Fee() *big.Int
+	Cost() *big.Int
+
+	EffectiveGasPrice(baseFee *big.Int) *big.Int
+	EffectiveFee(baseFee *big.Int) *big.Int
+	EffectiveCost(baseFee *big.Int) *big.Int
+}
+
+func NewTxDataFromTx(tx *ethtypes.Transaction) (TxData, error) {
+	var txData TxData
+	var err error
+	switch tx.Type() {
+	case ethtypes.DynamicFeeTxType:
+		txData, err = NewDynamicFeeTx(tx)
+	case ethtypes.AccessListTxType:
+		txData, err = NewAccessListTx(tx)
+	case ethtypes.BlobTxType:
+		txData, err = NewBlobTx(tx)
+	default:
+		txData, err = NewLegacyTx(tx)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return txData, nil
+}
+
+func rawSignatureValues(vBz, rBz, sBz []byte) (v, r, s *big.Int) {
+	if len(vBz) > 0 {
+		v = new(big.Int).SetBytes(vBz)
+	}
+	if len(rBz) > 0 {
+		r = new(big.Int).SetBytes(rBz)
+	}
+	if len(sBz) > 0 {
+		s = new(big.Int).SetBytes(sBz)
+	}
+	return v, r, s
+}
+
+// fee = gas limit * gas price
```

Comment:
could we resuse cosmos-sdk native impl here?

Reply:
there isn't an existing impl - it's spread across in https://github.com/sei-protocol/sei-cosmos/blob/main/x/auth/ante/validator_tx_fee.go#L25-L43